Add an encryption provider to encrypt data before storage

aewakefield opened the pull request:
This allows for events, side effects and snapshots to be encrypted before storing them.

aewakefield committed:
feat: add `EncryptionProvider`

This allows for events, side effects and snapshots to be encrypted
before storing them.

aewakefield committed:
chore: merge rewrite_traits into encryption

jdon commented on eventastic_postgres/src/lib.rs:
```diff
@@ -42,6 +44,9 @@ pub enum DbError {
     UnregisteredAggregate,
+    /// Failed to encrypt or decrypt data.
+    #[error("Encryption Error {0}")]
+    Encryption(anyhow::Error),
```
❓ Why does the encryption provider have an error type if we are always returning anyhow error? I think we should make this error generic and then use the both the encryption provider error and the picking error in this struct.

aewakefield replied:
I've made the encryption provider one generic but not the pickle one. If we make pickle generic here we're saying all implementations of pickle must return the same error type. Right now we could pickle different things in different ways with different errors. I think this potentially makes sense. We may want just the default serde_json for events but something specific for the aggregate. So maybe this more dynamic error type make sense for pickle. What do you think?

jdon commented on eventastic_postgres/src/reader_impl.rs:
```diff
@@ -4,86 +4,142 @@
 {
     let id = *id;
 
     Box::pin(async_stream::stream! {
```
⛏️ I think we can remove this Box::pin as the other functions pin the function before returning it.

jdon commented on eventastic_postgres/src/encryption.rs:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    fn max_batch_size(&self) -> usize {
+        1
```
It looks like we allocate a vec everytime we have a batch to decrypt, so it's probably best to have this larger than 1. What about 100 by default?

jdon commented on eventastic_postgres/src/transaction.rs:
```diff
@@ -222,6 +256,21 @@ where
+        let aggregate = cipher
+            .into_iter()
+            .next()
+            .expect("Must have encrypted snapshot");
```
I dislike these two checks being independent as we could easily remove the check and then start having panics. 

Can we do just one check and get the aggregate at the same time e.g

```suggestion
        let aggregate = cipher.into_iter().next().ok_or_else(|| {
            DbError::Encryption(anyhow!("Encrypting snapshot returned empty result"))
        })?;
```

aewakefield replied:
That change isn't quite the same, if the encryption provider returns more items then that won't catch it. So it's still two statements but it doesn't use `.expect` anymore.

jdon reviewed (commented):
Overall looks good, I think the main thing to change is the error type that is returned for pickling / encrypting so it's no longer anyhow.

Also looks like there are a couple of clippy lints to fix .

aewakefield committed:
chore: fix pr comments

aewakefield committed:
wip: side effect trait must be generic

mapokko committed:
fix: finish updating generic

mapokko committed:
fix: format

aewakefield committed:
chore: tidy up

jdon committed:
Remove table registry

Co-Authored-By: Alex Wakefield <1181370+aewakefield@users.noreply.github.com>
Co-Authored-By: Fabio Mirza <79943909+mapokko@users.noreply.github.com>

aewakefield committed:
chore: simplify the error message

When we get this error logged we end up with lots of different
messages due to the aggregate ID and version number. These should be
logged outside of the message to make grouping this easier.

jdon committed:
Merge pull request #17 from jdon/simplify_error_message

chore: simplify the error message

jdon committed:
Merge pull request #16 from jdon/side-effects-changes

Side effects changes

jdon committed:
Merge pull request #15 from jdon/side-effects

Generic SideEffectStorage
